syntax = "proto3";

package ai.stigmer.commons.apiresource.apiresourcekind;

import "ai/stigmer/commons/apiresource/apiresourcekind/api_resource_group.proto";
import "google/protobuf/descriptor.proto";

option java_outer_classname = "ApiResourceKindOuterClass";

// Enum value options for api-resource-kind
extend google.protobuf.EnumValueOptions {
  ApiResourceKindMeta kind_meta = 90101;
}

// API resource version
enum ApiResourceVersion {
  api_resource_version_unspecified = 0;
  v1 = 1;
}

// Resource tier defines the availability context of a resource
enum ResourceTier {
  TIER_UNSPECIFIED = 0;
  TIER_OPEN_SOURCE = 1; // Available in CLI local mode & Cloud
  TIER_CLOUD_ONLY = 2; // Hidden in CLI local mode; Available in Cloud
}

// Platform ID values - defines the singleton platform instance ID
// This enum is used to define the platform kind value wherever platform kind is required
enum PlatformIdValue {
  platform_id_value_unspecified = 0;

  // The Stigmer platform singleton instance
  // This value is used as the resource ID for platform-level authorization
  stigmer = 1;
}

// Metadata for api-resource-kind
message ApiResourceKindMeta {
  // API resource group
  ApiResourceGroup group = 1;
  // API resource version
  ApiResourceVersion version = 2;
  // Kind name
  string name = 3;
  // Display name
  string display_name = 4;
  // Prefix to be used for api-resource metadata.id field
  string id_prefix = 5;
  // Flag to indicate if the api-resource-kind is versioned
  bool is_versioned = 6;
  // Flag to indicate if the api-resource-kind is not searchable
  bool not_search_indexed = 7;
  // Defines availability context - whether resource is available in open source or cloud only
  ResourceTier tier = 8;
}

// API resource kinds for Stigmer
enum ApiResourceKind {
  api_resource_kind_unknown = 0;

  // Agentic - System-level audit trails and versioning
  api_resource_version = 1 [(kind_meta) = {
    group: agentic
    version: v1
    name: "ApiResourceVersion"
    display_name: "API Resource Version"
    id_prefix: "ver"
    is_versioned: false
    not_search_indexed: true
    tier: TIER_CLOUD_ONLY
  }];

  // IAM - Identity and Access Management
  iam_policy = 10 [(kind_meta) = {
    group: iam
    version: v1
    name: "IamPolicy"
    display_name: "IAM Policy"
    id_prefix: "iamp"
    is_versioned: false
    not_search_indexed: true
    tier: TIER_CLOUD_ONLY
  }];

  identity_account = 11 [(kind_meta) = {
    group: iam
    version: v1
    name: "IdentityAccount"
    display_name: "Identity Account"
    id_prefix: "ida"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_CLOUD_ONLY
  }];

  api_key = 12 [(kind_meta) = {
    group: iam
    version: v1
    name: "ApiKey"
    display_name: "API Key"
    id_prefix: "key"
    is_versioned: false
    not_search_indexed: true
    tier: TIER_CLOUD_ONLY
  }];

  // IAM - Credential and connection management
  credential = 20 [(kind_meta) = {
    group: iam
    version: v1
    name: "Credential"
    display_name: "Credential"
    id_prefix: "crd"
    is_versioned: false
    not_search_indexed: true
    tier: TIER_CLOUD_ONLY
  }];

  // Tenancy - Platform and organizational resources
  organization = 30 [(kind_meta) = {
    group: tenancy
    version: v1
    name: "Organization"
    display_name: "Organization"
    id_prefix: "org"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_CLOUD_ONLY
  }];

  platform = 31 [(kind_meta) = {
    group: tenancy
    version: v1
    name: "Platform"
    display_name: "Platform"
    id_prefix: "plt"
    is_versioned: false
    not_search_indexed: true
    tier: TIER_CLOUD_ONLY
  }];

  // Agentic - AI agent management
  agent = 40 [(kind_meta) = {
    group: agentic
    version: v1
    name: "Agent"
    display_name: "Agent"
    id_prefix: "agt"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  agent_execution = 41 [(kind_meta) = {
    group: agentic
    version: v1
    name: "AgentExecution"
    display_name: "Agent Execution"
    id_prefix: "aex"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  session = 42 [(kind_meta) = {
    group: agentic
    version: v1
    name: "Session"
    display_name: "Session"
    id_prefix: "ses"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  skill = 43 [(kind_meta) = {
    group: agentic
    version: v1
    name: "Skill"
    display_name: "Skill"
    id_prefix: "skl"
    is_versioned: true
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  mcp_server = 44 [(kind_meta) = {
    group: agentic
    version: v1
    name: "McpServer"
    display_name: "MCP Server"
    id_prefix: "mcp"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  agent_instance = 45 [(kind_meta) = {
    group: agentic
    version: v1
    name: "AgentInstance"
    display_name: "Agent Instance"
    id_prefix: "ain"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  // Agentic - Workflow orchestration
  workflow = 50 [(kind_meta) = {
    group: agentic
    version: v1
    name: "Workflow"
    display_name: "Workflow"
    id_prefix: "wfl"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  workflow_instance = 51 [(kind_meta) = {
    group: agentic
    version: v1
    name: "WorkflowInstance"
    display_name: "Workflow Instance"
    id_prefix: "win"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  workflow_execution = 52 [(kind_meta) = {
    group: agentic
    version: v1
    name: "WorkflowExecution"
    display_name: "Workflow Execution"
    id_prefix: "wex"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  environment = 53 [(kind_meta) = {
    group: agentic
    version: v1
    name: "Environment"
    display_name: "Environment"
    id_prefix: "env"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];

  execution_context = 54 [(kind_meta) = {
    group: agentic
    version: v1
    name: "ExecutionContext"
    display_name: "Execution Context"
    id_prefix: "exec"
    is_versioned: false
    not_search_indexed: false
    tier: TIER_OPEN_SOURCE
  }];
}

// Note: Extension field 'kind_meta' (90101) is used for metadata access in Java
